In [2]:
import pandas as pd
from sklearn.metrics import f1_score
from sklearn.metrics import classification_report

In [6]:
#loading the data
train_data = pd.read_csv('/home/rajesh/Downloads/train.csv')
test_data = pd.read_csv('/home/rajesh/Downloads/test.csv')
#print(train_data)

In [7]:
#feature removal that doesnt have any weight 
train_data = train_data.drop(['ID','Batch Enrolled','Grade','Sub Grade','Payment Plan'], axis=1)
test_data = test_data.drop(['ID','Batch Enrolled','Grade','Sub Grade','Payment Plan'], axis=1)

In [8]:
#train_data.head()
test_data.head()

Loan Amount  Funded Amount  Funded Amount Investor  Term  Interest Rate  \
0        17120          10365            16025.082690    59      12.163926   
1         7133          11650            12615.795600    59       6.564296   
2        25291          25825            11621.280830    59      14.729900   
3        30781           9664            15375.823510    59      10.523767   
4         8878           9419             7176.647582    58       9.997013   

  Employment Duration  Home Ownership Verification Status  \
0                RENT     76468.82190     Source Verified   
1            MORTGAGE     38079.01344     Source Verified   
2            MORTGAGE     51275.93268     Source Verified   
3                RENT     68867.98965            Verified   
4                 OWN     91556.85423            Verified   

                Loan Title  Debit to Income  ...  Total Received Late Fee  \
0         Home improvement        16.749219  ...                 0.088031   
1  Credit card refinancing        18.157975  ...                 0.041237   
2       Debt Consolidation        15.190011  ...                 0.021745   
3       Debt Consolidation        21.292550  ...                 0.092398   
4       Debt Consolidation         4.812117  ...                 0.010354   

   Recoveries  Collection Recovery Fee  Collection 12 months Medical  \
0    8.425776                 0.731797                             0   
1    6.157008                 0.992918                             0   
2    5.705077                 0.281580                             0   
3    2.469688                 0.959162                             0   
4    2.127835                 0.402315                             0   

   Application Type  Last week Pay  Accounts Delinquent  \
0        INDIVIDUAL            135                    0   
1        INDIVIDUAL             56                    0   
2        INDIVIDUAL              3                    0   
3        INDIVIDUAL             21                    0   
4        INDIVIDUAL            104                    0   

  Total Collection Amount  Total Current Balance  Total Revolving Credit Limit  
0                      24                 475442                          4364  
1                       1                  72412                          2573  
2                      26                 284825                         19676  
3                      32                  40842                          7226  
4                      33                  90825                         26145  

[5 rows x 29 columns]

In [9]:
#encoding categorical values to numerical values
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
train_data["Employment Duration"] = le.fit_transform(train_data["Employment Duration"])
train_data["Verification Status"] = le.fit_transform(train_data["Verification Status"])
train_data["Loan Title"] = le.fit_transform(train_data["Loan Title"])
train_data["Application Type"] = le.fit_transform(train_data["Application Type"])
train_data["Initial List Status"] = le.fit_transform(train_data["Initial List Status"])

print(train_data)

       Loan Amount  Funded Amount  Funded Amount Investor  Term  \
0            10000          32236             12329.36286    59   
1             3609          11940             12191.99692    59   
2            28276           9311             21603.22455    59   
3            11170           6954             17877.15585    59   
4            16890          13226             13539.92667    59   
...            ...            ...                     ...   ...   
67458        13601           6848             13175.28583    59   
67459         8323          11046             15637.46301    59   
67460        15897          32921             12329.45775    59   
67461        16567           4975             21353.68465    59   
67462        15353          29875             14207.44860    59   

       Interest Rate  Employment Duration  Home Ownership  \
0          11.135007                    0    176346.62670   
1          12.237563                    2     39833.92100   
2          1

In [10]:
#splitting the preprocessed data into fetures and label
label = train_data["Loan Status"]
features = train_data.drop(["Loan Status"], axis = 1)
print(features)

       Loan Amount  Funded Amount  Funded Amount Investor  Term  \
0            10000          32236             12329.36286    59   
1             3609          11940             12191.99692    59   
2            28276           9311             21603.22455    59   
3            11170           6954             17877.15585    59   
4            16890          13226             13539.92667    59   
...            ...            ...                     ...   ...   
67458        13601           6848             13175.28583    59   
67459         8323          11046             15637.46301    59   
67460        15897          32921             12329.45775    59   
67461        16567           4975             21353.68465    59   
67462        15353          29875             14207.44860    59   

       Interest Rate  Employment Duration  Home Ownership  \
0          11.135007                    0    176346.62670   
1          12.237563                    2     39833.92100   
2          1

In [26]:
#standardising the data
from sklearn.preprocessing import StandardScaler
std = StandardScaler(with_mean=False)
features =std.fit_transform(features)

In [ ]:
#splitting the data into train and test f
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(features, label, test_size=0.3, random_state=42)
print(X_train)

[[0.6912208  2.7800501  1.95454494 ... 0.05910986 1.1852289  0.3679862 ]
 [2.32581938 1.17642985 3.35286568 ... 0.07791754 4.64185225 2.69145789]
 [3.62675811 3.25140722 1.95092974 ... 0.06045326 0.4393586  1.20583968]
 ...
 [4.11266817 1.08183944 1.97271892 ... 4.80939277 2.61866647 0.31683053]
 [1.83142441 0.77966542 1.76350854 ... 0.04164558 1.5814752  1.36590436]
 [1.36845944 1.08515195 1.33755761 ... 0.04970601 2.02051014 0.57925437]]


In [28]:
#oversampling minority class by SMOte
from imblearn.over_sampling import SMOTE
from collections import Counter
counter = Counter(y_train)
print("before", counter)
smt = SMOTE()
x_train_sm, y_train_sm = smt.fit_resample(X_train,y_train)
counter = Counter(y_train_sm)
print("after", counter)
#print(x_train_sm)


before Counter({0: 42836, 1: 4388})


after Counter({0: 42836, 1: 42836})


In [11]:
#oversampling minority class by ADASYN
from imblearn.over_sampling import ADASYN
from collections import Counter
counter = Counter(y_train)
print("before", counter)
ada = ADASYN(random_state = 0)
x_train_ada, y_train_ada = ada.fit_resample(X_train,y_train)
counter = Counter(y_train_ada)
print("after", counter)
#x_train_sm.isnull().values.any()

before Counter({0: 42836, 1: 4388})


after Counter({1: 44306, 0: 42836})


In [12]:
#oversampling and undersampling by SMOTE+TOmek
from imblearn.combine import SMOTETomek
from collections import Counter
counter = Counter(y_train)
print("before", counter)
smtok = SMOTETomek()
x_train_smtok, y_train_smtok = smtok.fit_resample(X_train,y_train)
counter = Counter(y_train_smtok)
print("after", counter)
#x_train_sm.isnull().values.any()

before Counter({0: 42836, 1: 4388})


after Counter({0: 42830, 1: 42830})


In [13]:
#Oversampling and undersampling BY SMOTE+ENN
from imblearn.combine import SMOTEENN
from collections import Counter
counter = Counter(y_train)
print("before", counter)
smtnn = SMOTEENN()
x_train_smtnn, y_train_smtnn = smtnn.fit_resample(X_train,y_train)
counter = Counter(y_train_smtnn)
print("after", counter)
#x_train_sm.isnull().values.any()

before Counter({0: 42836, 1: 4388})


after Counter({1: 42775, 0: 20326})


In [20]:
#Logistic Regression on SMote balanced data
from sklearn.linear_model import LogisticRegression#Smote
from sklearn.metrics import accuracy_score
from sklearn.metrics import average_precision_score
clf = LogisticRegression(random_state=0)
clf.fit(x_train_sm, y_train_sm)

y_pred_log = clf.predict(X_test)
prec_log = average_precision_score(y_test, y_pred_log)
acc_log = accuracy_score(y_test, y_pred_log)
classr_log =classification_report(y_test, y_pred_log)

print(y_pred_log)
print(prec_log)
print(classr_log)
print(acc_log)


[1 0 1 ... 0 0 0]
0.09136687002184882
              precision    recall  f1-score   support

           0       0.91      0.50      0.64     18386
           1       0.09      0.50      0.15      1853

    accuracy                           0.50     20239
   macro avg       0.50      0.50      0.40     20239
weighted avg       0.83      0.50      0.60     20239

0.4981471416571965


In [15]:
#Decision tree on Smote balanced data
from sklearn.tree import DecisionTreeClassifier

DT = DecisionTreeClassifier()
DT.fit(x_train_sm, y_train_sm)

y_pred_DT = DT.predict(X_test)
prec_DT = average_precision_score(y_test, y_pred_DT)
acc_DT = accuracy_score(y_test, y_pred_DT)
classr_DT =classification_report(y_test, y_pred_DT)

print(prec_DT)
print(acc_DT)
print(classr_DT)


0.09158394161683722
0.7963338109590394
              precision    recall  f1-score   support

           0       0.91      0.86      0.89     18386
           1       0.09      0.14      0.11      1853

    accuracy                           0.80     20239
   macro avg       0.50      0.50      0.50     20239
weighted avg       0.83      0.80      0.81     20239



In [ ]:
#decision tree hyper parameter tuning on smote data
#Define a param_grid dictionary for the given hyper parameters
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier


criterion=["gini", "entropy", "log_loss"]
splitter=["best", "random"]
max_depth=[100,150,200]
min_samples_split=[2,5]
min_samples_leaf=[1,5]
max_leaf_nodes=[100,10000,1e6]
max_features=["auto", "sqrt", "log2"]
param_grid = {"criterion": criterion, "splitter": splitter, "max_features": max_features,
              "max_depth": max_depth, "min_samples_split": min_samples_split,
              "min_samples_leaf": min_samples_leaf, "max_leaf_nodes": max_leaf_nodes}

#call decision tree classifier
dtc = DecisionTreeClassifier()

#apply gridsearchcv for hyper parameter tuning
gridSearch = GridSearchCV(estimator=dtc,param_grid=param_grid,cv=5, verbose=2)

#fit the model
gridSearch.fit(x_train_sm, y_train_sm)

#output of above code(it was ran in another computer)
Fitting 5 folds for each of 648 candidates, totalling 3240 fits
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=2, splitter=best; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=2, splitter=best; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=2, splitter=best; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=2, splitter=best; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=2, splitter=best; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=2, splitter=random; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=2, splitter=random; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=2, splitter=random; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=2, splitter=random; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=2, splitter=random; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=5, splitter=best; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=5, splitter=best; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=5, splitter=best; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=5, splitter=best; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=5, splitter=best; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=5, splitter=random; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=5, splitter=random; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=5, splitter=random; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=5, splitter=random; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=1, min_samples_split=5, splitter=random; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=5, min_samples_split=2, splitter=best; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=5, min_samples_split=2, splitter=best; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=5, min_samples_split=2, splitter=best; total time=   0.0s
[CV] END criterion=gini, max_depth=100, max_features=auto, max_leaf_nodes=100, min_samples_leaf=5, min_samples_split=2, splitter=best; total time=   0.0s
...
[CV] END criterion=log_loss, max_depth=200, max_features=log2, max_leaf_nodes=1000000.0, min_samples_leaf=5, min_samples_split=5, splitter=random; total time=   0.0s
[CV] END criterion=log_loss, max_depth=200, max_features=log2, max_leaf_nodes=1000000.0, min_samples_leaf=5, min_samples_split=5, splitter=random; total time=   0.0s
[CV] END criterion=log_loss, max_depth=200, max_features=log2, max_leaf_nodes=1000000.0, min_samples_leaf=5, min_samples_split=5, splitter=random; total time=   0.0s
[CV] END criterion=log_loss, max_depth=200, max_features=log2, max_leaf_nodes=1000000.0, min_samples_leaf=5, min_samples_split=5, splitter=random; total time=   0.0s


In [ ]:
#print the best parameters
print(gridSearch.best_estimator_)


OUTPUT
DecisionTreeClassifier(criterion='log_loss', max_depth=200, max_features='sqrt',
                       max_leaf_nodes=10000, min_samples_split=5)

In [ ]:
#fitting the model with best parameters
dtcFinal=DecisionTreeClassifier(criterion='log_loss', max_depth=200, max_features='sqrt',
                       max_leaf_nodes=10000, min_samples_split=5)

dtcFinal.fit(x_train_sm, y_train_sm)

#Predict the labels for test data
y_test_pred_dtcf = dtcFinal.predict(X_test)

acc_dt = accuracy_score(y_test, y_test_pred_dtcf)
classr_dt =classification_report(y_test, y_test_pred_dtcf)
f_score_dt = f1_score(y_test, y_test_pred_dtcf, average = None)
roc_auc_SCORE_dt = roc_auc_score(y_test, y_test_pred_dtcf)
print(acc_dt)
print(classr_dt)
print(f_score_dt)
print(roc_auc_SCORE_dt)

OUTPUT
0.7901082069272197
              precision    recall  f1-score   support

           0       0.91      0.85      0.88     18386
           1       0.11      0.18      0.13      1853

    accuracy                           0.79     20239
   macro avg       0.51      0.51      0.51     20239
weighted avg       0.84      0.79      0.81     20239

[0.88058695 0.13376835]
0.5144542185215775


In [16]:
#Random Forest on Smote balanced data
from sklearn.ensemble import RandomForestClassifier
RFC = RandomForestClassifier()
RFC.fit(x_train_sm, y_train_sm)

y_pred_RFC = RFC.predict(X_test)
prec_RFC = average_precision_score(y_test, y_pred_RFC)
acc_RFC = accuracy_score(y_test, y_pred_RFC)
classr_RFC =classification_report(y_test, y_pred_RFC)
print(prec_RFC)
print(acc_RFC)
print(classr_RFC)


0.09209149352057093
0.9076535401946737
              precision    recall  f1-score   support

           0       0.91      1.00      0.95     18386
           1       0.23      0.00      0.01      1853

    accuracy                           0.91     20239
   macro avg       0.57      0.50      0.48     20239
weighted avg       0.85      0.91      0.87     20239



In [17]:
#hyperparemeter tuning of random forest on smote data
from sklearn.model_selection import RandomizedSearchCV
n_estimators=[20,50,100,150,200,300]
max_features=['sqrt', 'log2', None, 'auto', 20, 30, 40]
max_depth=[None, 10,100,1000,10000,1e5]
min_samples_split=[1,5,7,9,11,13]
min_samples_leaf=[2,3,5,7,9,11]
max_leaf_nodes=[100,1000,10000,1e5,1e6]
bootstrap=[True, False]

param_distributions={"n_estimators": n_estimators, "max_features": max_features,
                     "max_depth": max_depth, "min_samples_split": min_samples_split,
                     "min_samples_leaf": min_samples_leaf, "max_leaf_nodes": max_leaf_nodes,
                     "bootstrap": bootstrap} 
#call random forest classifier
RFC = RandomForestClassifier()
#Apply randomised search cv for hyper parameter tuning
Randsearch=RandomizedSearchCV(estimator=RFC, param_distributions=param_distributions,cv=5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               )
#fit the model
Randsearch.fit(x_train_sm, y_train_sm)

/home/rajesh/anaconda3/lib/python3.11/site-packages/sklearn/model_selection/_validation.py:425: FitFailedWarning: 
30 fits failed out of a total of 50.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
15 fits failed with the following error:
Traceback (most recent call last):
  File "/home/rajesh/anaconda3/lib/python3.11/site-packages/sklearn/model_selection/_validation.py", line 732, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/home/rajesh/anaconda3/lib/python3.11/site-packages/sklearn/base.py", line 1144, in wrapper
    estimator._validate_params()
  File "/home/rajesh/anaconda3/lib/python3.11/site-packages/sklearn/base.py", line 637, in _validate_params
    validate_parameter_constraints(
  File "/home/r

RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(),
                   param_distributions={'bootstrap': [True, False],
                                        'max_depth': [None, 10, 100, 1000,
                                                      10000, 100000.0],
                                        'max_features': ['sqrt', 'log2', None,
                                                         'auto', 20, 30, 40],
                                        'max_leaf_nodes': [100, 1000, 10000,
                                                           100000.0,
                                                           1000000.0],
                                        'min_samples_leaf': [2, 3, 5, 7, 9, 11],
                                        'min_samples_split': [1, 5, 7, 9, 11,
                                                              13],
                                        'n_estimators': [20, 50, 100, 150, 200,
                                                         300]})

In [18]:
#finifng the best parameters

Best_parameter = Randsearch.best_estimator_
print(Best_parameter)




RandomForestClassifier(max_depth=1000, max_features=20, max_leaf_nodes=1000,
                       min_samples_leaf=2, min_samples_split=9,
                       n_estimators=300)


In [19]:
#fitting the model with best parametrs
RFCFinal = RandomForestClassifier(max_depth=1000, max_features=20, max_leaf_nodes=1000,
                       min_samples_leaf=2, min_samples_split=9,
                       n_estimators=300)
RFCFinal.fit(x_train_sm, y_train_sm)



y_train_pred_RFCF = RFCFinal.predict(X_test)
prec_RFCF = average_precision_score(y_test, y_train_pred_RFCF )
acc_RFCF = accuracy_score(y_test, y_train_pred_RFCF )
classr_RFCF =classification_report(y_test, y_train_pred_RFCF )


print(prec_RFCF)
print(classr_RFCF)
#Predict the labels for test data

0.09270172316413602
              precision    recall  f1-score   support

           0       0.91      1.00      0.95     18386
           1       0.19      0.01      0.02      1853

    accuracy                           0.91     20239
   macro avg       0.55      0.50      0.49     20239
weighted avg       0.84      0.91      0.87     20239



In [21]:
print(acc_RFCF)

0.9051336528484609


In [10]:
#svm using smote data, it will take  a bit long
from sklearn.svm import SVC
svm_model = SVC(kernel='linear', random_state=42)
svm_model.fit(x_train_sm, y_train_sm) 

y_pred_svm = svm_model.predict(X_test)
prec_svm = average_precision_score(y_test, y_pred_svm)
acc_svm = accuracy_score(y_test, y_pred_svm)
classr_svm =classification_report(y_test, y_pred_svm)

print(y_pred_svm)
print(prec_svm)
print(classr_svm)

In [ ]:
#XGboost using on smote data
from sklearn.ensemble import GradientBoostingClassifier

xgb= GradientBoostingClassifier(n_estimators=100, learning_rate=0.1,max_depth=3, random_state=0)
xgb.fit(x_train_sm, y_train_sm)

y_pred_xgb = xgb.predict(X_test)
prec_xgb = average_precision_score(y_test, y_pred_xgb)
acc_xgb = accuracy_score(y_test, y_pred_xgb)
classr_xgb =classification_report(y_test, y_pred_xgb)

print(prec_xgb)
print(acc_xgb)
print(classr_xgb)


0.09214253842609857
0.9079499975295222
              precision    recall  f1-score   support

           0       0.91      1.00      0.95     18386
           1       0.27      0.00      0.01      1853

    accuracy                           0.91     20239
   macro avg       0.59      0.50      0.48     20239
weighted avg       0.85      0.91      0.87     20239



In [ ]:
#Logistic regression using Adasyn data
clf2 = LogisticRegression(random_state=0)#ADASYN
clf2.fit(x_train_ada, y_train_ada)

y_pred_log2 = clf2.predict(X_test)
acc_log2 = accuracy_score(y_test, y_pred_log2)
classr_log2 =classification_report(y_test, y_pred_log2)


print(classr_log2)
print(acc_log2)
#print(y_pred2)

              precision    recall  f1-score   support

           0       0.91      0.41      0.56     18386
           1       0.09      0.58      0.16      1853

    accuracy                           0.42     20239
   macro avg       0.50      0.49      0.36     20239
weighted avg       0.83      0.42      0.52     20239

0.4229457977172785


In [ ]:
#deciosontree using adasyn data
DT2 = DecisionTreeClassifier()
DT2.fit(x_train_ada, y_train_ada)

y_pred_DT2 = DT2.predict(X_test)
prec_DT2 = average_precision_score(y_test, y_pred_DT2)
acc_DT2 = accuracy_score(y_test, y_pred_DT2)
classr_DT2 =classification_report(y_test, y_pred_DT2)

print(prec_DT2)
print(acc_DT2)
print(classr_DT2)


0.09294785000941717
0.7970749542961609
              precision    recall  f1-score   support

           0       0.91      0.86      0.89     18386
           1       0.10      0.15      0.12      1853

    accuracy                           0.80     20239
   macro avg       0.51      0.51      0.50     20239
weighted avg       0.84      0.80      0.82     20239



In [ ]:
#Randomforest using Adasyn data
RFC2 = RandomForestClassifier()
RFC2.fit(x_train_ada, y_train_ada)

y_pred_RFC2 = RFC2.predict(X_test)
prec_RFC2 = average_precision_score(y_test, y_pred_RFC2)
acc_RFC2 = accuracy_score(y_test, y_pred_RFC2)
classr_RFC2 =classification_report(y_test, y_pred_RFC2)

print(prec_RFC2)
print(acc_RFC2)
print(classr_RFC2)


0.09196557159216653
0.9074064924156332
              precision    recall  f1-score   support

           0       0.91      1.00      0.95     18386
           1       0.20      0.00      0.01      1853

    accuracy                           0.91     20239
   macro avg       0.55      0.50      0.48     20239
weighted avg       0.84      0.91      0.87     20239



In [ ]:
#XGBoost on ADASYn data
xgb2= GradientBoostingClassifier(n_estimators=100, learning_rate=1.0,max_depth=5, random_state=0)
xgb2.fit(x_train_ada, y_train_ada)

y_pred_xgb2 = xgb2.predict(X_test)
prec_xgb2 = average_precision_score(y_test, y_pred_xgb2)
acc_xgb2 = accuracy_score(y_test, y_pred_xgb2)
classr_xgb2 =classification_report(y_test, y_pred_xgb2)

print(prec_xgb2)
print(acc_xgb2)
print(classr_xgb2)


0.09268700911548443
0.8831958100696675
              precision    recall  f1-score   support

           0       0.91      0.97      0.94     18386
           1       0.12      0.04      0.06      1853

    accuracy                           0.88     20239
   macro avg       0.51      0.51      0.50     20239
weighted avg       0.84      0.88      0.86     20239



In [ ]:
#Logistic regression using SMOTETEk data
clf3 = LogisticRegression(random_state=0) #SMOTETomek
clf3.fit(x_train_smtok, y_train_smtok)

y_pred_log3 = clf3.predict(X_test)
acc_log3 = accuracy_score(y_test, y_pred_log3)
classr_log3 =classification_report(y_test, y_pred_log3)


print(classr_log3)

print(acc_log3)
#print(y_pred2)

              precision    recall  f1-score   support

           0       0.91      0.49      0.64     18386
           1       0.09      0.51      0.15      1853

    accuracy                           0.50     20239
   macro avg       0.50      0.50      0.40     20239
weighted avg       0.83      0.50      0.60     20239

0.4954296160877514


In [ ]:
#decision tree using SMOTETEK data
DT3 = DecisionTreeClassifier()
DT3.fit(x_train_smtok, y_train_smtok)

y_pred_DT3 = DT3.predict(X_test)
prec_DT3 = average_precision_score(y_test, y_pred_DT3)
acc_DT3 = accuracy_score(y_test, y_pred_DT3)
classr_DT3 =classification_report(y_test, y_pred_DT3)

print(prec_DT3)
print(acc_DT3)
print(classr_DT3)


0.09084915215773971
0.7926775038292406
              precision    recall  f1-score   support

           0       0.91      0.86      0.88     18386
           1       0.09      0.13      0.10      1853

    accuracy                           0.79     20239
   macro avg       0.50      0.50      0.49     20239
weighted avg       0.83      0.79      0.81     20239



In [ ]:
#Randomforest using SMOtetek data
RFC3 = RandomForestClassifier()
RFC3.fit(x_train_smtok, y_train_smtok)

y_pred_RFC3 = RFC3.predict(X_test)
prec_RFC3 = average_precision_score(y_test, y_pred_RFC3)
acc_RFC3 = accuracy_score(y_test, y_pred_RFC3)
classr_RFC3 =classification_report(y_test, y_pred_RFC3)

print(prec_RFC3)
print(acc_RFC3)
print(classr_RFC3)


0.09218943279821878
0.907801768862098
              precision    recall  f1-score   support

           0       0.91      1.00      0.95     18386
           1       0.26      0.00      0.01      1853

    accuracy                           0.91     20239
   macro avg       0.58      0.50      0.48     20239
weighted avg       0.85      0.91      0.87     20239



In [ ]:
#XGboost using smtotek data
xgb3= GradientBoostingClassifier(n_estimators=100, learning_rate=1.0,max_depth=5, random_state=0)
xgb3.fit(x_train_smtok, y_train_smtok)

y_pred_xgb3 = xgb3.predict(X_test)
prec_xgb3 = average_precision_score(y_test, y_pred_xgb3)
acc_xgb3 = accuracy_score(y_test, y_pred_xgb3)
classr_xgb3 =classification_report(y_test, y_pred_xgb3)

print(prec_xgb3)
print(acc_xgb3)
print(classr_xgb3)


0.09231390009440477
0.8852216018577993
              precision    recall  f1-score   support

           0       0.91      0.97      0.94     18386
           1       0.11      0.04      0.06      1853

    accuracy                           0.89     20239
   macro avg       0.51      0.50      0.50     20239
weighted avg       0.84      0.89      0.86     20239



In [ ]:
#Logistic regression using SMOTEENN data
clf4 = LogisticRegression(random_state=0) #SMOTEENN
clf4.fit(x_train_smtnn, y_train_smtnn)

y_pred_log4 = clf4.predict(X_test)
acc_log4 = accuracy_score(y_test, y_pred_log4)
classr_log4 =classification_report(y_test, y_pred_log4)


print(classr_log4)

print(acc_log4)
#print(y_pred2)

              precision    recall  f1-score   support

           0       0.90      0.00      0.01     18386
           1       0.09      1.00      0.17      1853

    accuracy                           0.09     20239
   macro avg       0.50      0.50      0.09     20239
weighted avg       0.83      0.09      0.02     20239

0.09392756559118534


In [ ]:
#Decisiontree using SMOTEENN data
DT4 = DecisionTreeClassifier()
DT4.fit(x_train_smtnn, y_train_smtnn)

y_pred_DT4 = DT4.predict(X_test)
prec_DT4 = average_precision_score(y_test, y_pred_DT4)
acc_DT4 = accuracy_score(y_test, y_pred_DT4)
classr_DT4 =classification_report(y_test, y_pred_DT4)

print(prec_DT4)
print(acc_DT4)
print(classr_DT4)


0.09090282037931836
0.7088788971787143
              precision    recall  f1-score   support

           0       0.91      0.76      0.83     18386
           1       0.09      0.24      0.13      1853

    accuracy                           0.71     20239
   macro avg       0.50      0.50      0.48     20239
weighted avg       0.83      0.71      0.76     20239



In [ ]:
#Randomforest using SMOTEENN data
RFC4 = RandomForestClassifier()
RFC4.fit(x_train_smtnn, y_train_smtnn)

y_pred_RFC4 = RFC4.predict(X_test)
prec_RFC4 = average_precision_score(y_test, y_pred_RFC4)
acc_RFC4 = accuracy_score(y_test, y_pred_RFC4)
classr_RFC4 =classification_report(y_test, y_pred_RFC4)

print(prec_RFC4)
print(acc_RFC4)
print(classr_RFC4)


0.09234492698133938
0.8850239636345669
              precision    recall  f1-score   support

           0       0.91      0.97      0.94     18386
           1       0.11      0.04      0.06      1853

    accuracy                           0.89     20239
   macro avg       0.51      0.50      0.50     20239
weighted avg       0.84      0.89      0.86     20239



In [ ]:
#logistic using raw data without any sampling technique
clf5 = LogisticRegression(random_state=0) # NO technique
clf5.fit(X_train, y_train)

y_pred_log5 = clf5.predict(X_test)
acc_log5 = accuracy_score(y_test, y_pred_log5)
classr_log5 =classification_report(y_test, y_pred_log5)


print(classr_log5)
print(acc_log5)
#print(y_pred2)

              precision    recall  f1-score   support

           0       0.91      1.00      0.95     18386
           1       0.00      0.00      0.00      1853

    accuracy                           0.91     20239
   macro avg       0.45      0.50      0.48     20239
weighted avg       0.83      0.91      0.86     20239

0.9084440930876031


/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
#xgboost using raw data without sampling techniques
xgbl= GradientBoostingClassifier(n_estimators=100, learning_rate=0.1,max_depth=3, random_state=0)
xgbl.fit(X_train, y_train)

y_pred_xgbl = xgbl.predict(X_test)
prec_xgbl = average_precision_score(y_test, y_pred_xgbl)
acc_xgbl = accuracy_score(y_test, y_pred_xgbl)
classr_xgbl =classification_report(y_test, y_pred_xgbl)

print(prec_xgbl)
print(acc_xgbl)
print(classr_xgbl)


0.09155590691239686
0.9083946835317951
              precision    recall  f1-score   support

           0       0.91      1.00      0.95     18386
           1       0.00      0.00      0.00      1853

    accuracy                           0.91     20239
   macro avg       0.45      0.50      0.48     20239
weighted avg       0.83      0.91      0.86     20239



In [ ]:
#inbuilt sample balancing techniques with RFC
from imblearn.ensemble import BalancedRandomForestClassifier

blRFC = BalancedRandomForestClassifier()
blRFC.fit(X_train,y_train)
y_pred_blRFC = blRFC.predict(X_test)
prec_blRFC = average_precision_score(y_test, y_pred_blRFC)
acc_blRFC = accuracy_score(y_test, y_pred_blRFC)
classr_blRFC =classification_report(y_test, y_pred_blRFC)

print(prec_blRFC)
print(acc_blRFC)
print(classr_blRFC)


In [ ]:
#under sampling the data  using K-means
from imblearn.under_sampling import ClusterCentroids
from collections import Counter
undersampler = ClusterCentroids()
X_smote, y_smote = undersampler.fit_resample(X_train, y_train)
counter = Counter(y_train_sm)
print("after", counter)

/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


after Counter({0: 42836, 1: 42836})


In [ ]:
#logistic regression using K_means data
clf6 = LogisticRegression(random_state=0) # knn
clf6.fit(X_smote, y_smote)

y_pred_log6 = clf6.predict(X_test)
acc_log6 = accuracy_score(y_test, y_pred_log6)
classr_log6 =classification_report(y_test, y_pred_log6)


print(classr_log6)
print(acc_log6)

              precision    recall  f1-score   support

           0       0.91      0.28      0.43     18386
           1       0.09      0.72      0.16      1853

    accuracy                           0.32     20239
   macro avg       0.50      0.50      0.30     20239
weighted avg       0.83      0.32      0.41     20239

0.3236820000988191


In [ ]:
#decision tree using K-means  data
DT5 = DecisionTreeClassifier()
DT5.fit(X_smote, y_smote)

y_pred_DT5 = DT5.predict(X_test)
prec_DT5 = average_precision_score(y_test, y_pred_DT5)
acc_DT5 = accuracy_score(y_test, y_pred_DT5)
classr_DT5 =classification_report(y_test, y_pred_DT5)

print(prec_DT5)
print(acc_DT5)
print(classr_DT5)

0.09137494895941631
0.1684865853055981
              precision    recall  f1-score   support

           0       0.91      0.09      0.17     18386
           1       0.09      0.90      0.17      1853

    accuracy                           0.17     20239
   macro avg       0.50      0.50      0.17     20239
weighted avg       0.83      0.17      0.17     20239



In [ ]:
#Randomforest using K-means  data
RFC5 = RandomForestClassifier()
RFC5.fit(X_smote, y_smote)

y_pred_RFC5 = RFC5.predict(X_test)
prec_RFC5 = average_precision_score(y_test, y_pred_RFC5)
acc_RFC5 = accuracy_score(y_test, y_pred_RFC5)
classr_RFC5 =classification_report(y_test, y_pred_RFC5)

print(prec_RFC5)
print(acc_RFC5)
print(classr_RFC5)

0.09137856495811454
0.10504471564800633
              precision    recall  f1-score   support

           0       0.90      0.02      0.03     18386
           1       0.09      0.98      0.17      1853

    accuracy                           0.11     20239
   macro avg       0.49      0.50      0.10     20239
weighted avg       0.82      0.11      0.05     20239



In [12]:
# import library fro random under sampler
from imblearn.under_sampling import RandomUnderSampler
from collections import Counter
rus = RandomUnderSampler(random_state=42, replacement=True)# fit predictor and target variable
x_imbln_rus, y_imbln_rus = rus.fit_resample(X_train, y_train)

print('original dataset shape:', Counter(y_train))
print('Resample dataset shape', Counter(y_imbln_rus))

original dataset shape: Counter({0: 42836, 1: 4388})
Resample dataset shape Counter({0: 4388, 1: 4388})


In [30]:
# Logistic Regression ON DATA BY RANDOM UNDER SAMPLING

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import average_precision_score

clf = LogisticRegression()
clf.fit(x_imbln_rus, y_imbln_rus)

y_pred = clf.predict(X_test)
acc = accuracy_score(y_test, y_pred)
classr =classification_report(y_test, y_pred)
f_score = f1_score(y_test, y_pred, average = None)

print(acc)
print(classr)
print(f_score)


/home/rajesh/anaconda3/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/home/rajesh/anaconda3/lib/python3.11/site-packages/sklearn/base.py:464: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


0.5354513562923069
              precision    recall  f1-score   support

           0       0.91      0.54      0.68     18386
           1       0.09      0.46      0.15      1853

    accuracy                           0.54     20239
   macro avg       0.50      0.50      0.42     20239
weighted avg       0.83      0.54      0.63     20239

[0.67974658 0.1544964 ]


In [29]:
#decison tree usng Random sampler data
from sklearn.tree import DecisionTreeClassifier
DT6 = DecisionTreeClassifier()
DT6.fit(x_imbln_rus, y_imbln_rus)

y_pred_DT6 = DT6.predict(X_test)
prec_DT6 = average_precision_score(y_test, y_pred_DT6)
acc_DT6 = accuracy_score(y_test, y_pred_DT6)
classr_DT6 =classification_report(y_test, y_pred_DT6)

print(prec_DT6)
print(acc_DT6)
print(classr_DT6)

0.09166400451564019
0.4775927664410297
              precision    recall  f1-score   support

           0       0.91      0.47      0.62     18386
           1       0.09      0.53      0.16      1853

    accuracy                           0.48     20239
   macro avg       0.50      0.50      0.39     20239
weighted avg       0.83      0.48      0.58     20239



/home/rajesh/anaconda3/lib/python3.11/site-packages/sklearn/base.py:464: UserWarning: X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names
  warnings.warn(


In [19]:
#randomforest using radnom sampler data
from sklearn.ensemble import RandomForestClassifier
RFC6 = RandomForestClassifier()
RFC6.fit(x_imbln_rus, y_imbln_rus)

y_pred_RFC6 = RFC6.predict(X_test)
prec_RFC6 = average_precision_score(y_test, y_pred_RFC6)
acc_RFC6 = accuracy_score(y_test, y_pred_RFC6)
classr_RFC6 =classification_report(y_test, y_pred_RFC6)

print(prec_RFC6)
print(acc_RFC6)
print(classr_RFC6)

0.09517011690507998
0.5118335886160383
              precision    recall  f1-score   support

           0       0.92      0.51      0.65     18386
           1       0.10      0.53      0.17      1853

    accuracy                           0.51     20239
   macro avg       0.51      0.52      0.41     20239
weighted avg       0.84      0.51      0.61     20239



In [21]:
# IMPLEMENTING UNDERSAMPLING USING TOMEK LINKS
from imblearn.under_sampling import TomekLinks
from collections import Counter

counter = Counter(y_train)
print("before", counter)

tl = TomekLinks(sampling_strategy = 'majority')
x_train_tomek_links, y_train_tomek_links = tl.fit_resample(X_train,y_train)
counter = Counter(y_train_tomek_links)
print("after", counter)

before Counter({0: 42836, 1: 4388})
after Counter({0: 40950, 1: 4388})


In [23]:
# Logistic Regression ON NEW DATA BY TOMEK LINKS

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import average_precision_score

clf = LogisticRegression()
clf.fit(x_train_tomek_links, y_train_tomek_links)

y_pred = clf.predict(X_test)
acc = accuracy_score(y_test, y_pred)
classr =classification_report(y_test, y_pred)
f_score = f1_score(y_test, y_pred, average = None)

print(acc)
print(classr)
print(f_score)


0.9084440930876031
              precision    recall  f1-score   support

           0       0.91      1.00      0.95     18386
           1       0.00      0.00      0.00      1853

    accuracy                           0.91     20239
   macro avg       0.45      0.50      0.48     20239
weighted avg       0.83      0.91      0.86     20239

[0.95202589 0.        ]


/home/rajesh/anaconda3/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/home/rajesh/anaconda3/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/rajesh/anaconda3/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1469: UndefinedMetricWarning: Precision and F-score are ill-define

In [31]:
#decion tree using tomek data
DT7 = DecisionTreeClassifier()
DT7.fit(x_train_tomek_links, y_train_tomek_links)

y_pred_DT7 = DT7.predict(X_test)
prec_DT7 = average_precision_score(y_test, y_pred_DT7)
acc_DT7 = accuracy_score(y_test, y_pred_DT7)
classr_DT7 =classification_report(y_test, y_pred_DT7)

print(prec_DT7)
print(acc_DT7)
print(classr_DT7)

0.09191406952052486
0.3943376649043925
              precision    recall  f1-score   support

           0       0.91      0.37      0.53     18386
           1       0.09      0.63      0.16      1853

    accuracy                           0.39     20239
   macro avg       0.50      0.50      0.34     20239
weighted avg       0.83      0.39      0.49     20239



/home/rajesh/anaconda3/lib/python3.11/site-packages/sklearn/base.py:464: UserWarning: X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names
  warnings.warn(


In [32]:
#random forest using tomek data
RFC7 = RandomForestClassifier()
RFC7.fit(x_train_tomek_links, y_train_tomek_links)

y_pred_RFC7 = RFC7.predict(X_test)
prec_RFC7 = average_precision_score(y_test, y_pred_RFC7)
acc_RFC7 = accuracy_score(y_test, y_pred_RFC7)
classr_RFC7 =classification_report(y_test, y_pred_RFC7)

print(prec_RFC7)
print(acc_RFC7)
print(classr_RFC7)

/home/rajesh/anaconda3/lib/python3.11/site-packages/sklearn/base.py:464: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


0.0917359003553999
0.10321656208310687
              precision    recall  f1-score   support

           0       0.92      0.01      0.03     18386
           1       0.09      0.99      0.17      1853

    accuracy                           0.10     20239
   macro avg       0.51      0.50      0.10     20239
weighted avg       0.85      0.10      0.04     20239



In [ ]:
#XGboost using K-means data
xgb4= GradientBoostingClassifier(n_estimators=100, learning_rate=1.0,max_depth=5, random_state=0)
xgb4.fit(X_smote, y_smote)

y_pred_xgb4 = xgb4.predict(X_test)
prec_xgb4 = average_precision_score(y_test, y_pred_xgb4)
acc_xgb4 = accuracy_score(y_test, y_pred_xgb4)
classr_xgb4 =classification_report(y_test, y_pred_xgb4)

print(prec_xgb4)
print(acc_xgb4)
print(classr_xgb4)

0.09153769609427756
0.11957112505558574
              precision    recall  f1-score   support

           0       0.91      0.03      0.07     18386
           1       0.09      0.97      0.17      1853

    accuracy                           0.12     20239
   macro avg       0.50      0.50      0.12     20239
weighted avg       0.83      0.12      0.08     20239



In [ ]:
#RFECV on the raw data
from sklearn.feature_selection import RFECV
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import StratifiedKFold

In [ ]:
#feauture selection using RFEC with Randomforest as estimator
estimator = RandomForestClassifier()
selector = RFECV(estimator=estimator,scoring='accuracy', cv=StratifiedKFold(n_splits=5))
selector.fit_transform(X_train, y_train)
s_fetaures = X_train.columns[selector.support_]
print(s_fetaures)

OUTPUT
Index(['Loan Amount', 'Funded Amount', 'Funded Amount Investor',
       'Interest Rate', 'Home Ownership', 'Debit to Income',
       'Revolving Balance', 'Revolving Utilities', 'Total Received Interest',
       'Total Received Late Fee', 'Recoveries', 'Collection Recovery Fee',
       'Last week Pay', 'Total Current Balance',
       'Total Revolving Credit Limit'],
      dtype='object')

In [ ]:
# Assuming 'selector' is your RFECV object
selected_features = selector.get_support()

# Filter the original dataset based on selected features
data_train_selected = X_train.iloc[:, selected_features]
data_test_selected = X_test.iloc[:, selected_features]
data_test_selected.head
data_train_selected.head

In [ ]:
#oversampling technique ON RFECV selected data
from imblearn.over_sampling import SMOTE
from collections import Counter
counter = Counter(y_train)
print("before", counter)
smt = SMOTE()
x_train_smR, y_train_smR = smt.fit_resample(data_train_selected,y_train)
counter = Counter(y_train_sm)
print("after", counter)

OUTPUT
before Counter({0: 42836, 1: 4388})
after Counter({0: 42836, 1: 42836})

In [ ]:
#Logistic Regression usin SMote+RFECV technique
from sklearn.linear_model import LogisticRegression#Smote
from sklearn.metrics import accuracy_score
from sklearn.metrics import average_precision_score
clfR = LogisticRegression(random_state=0)
clfR.fit(x_train_smR, y_train_smR)
y_pred_logR = clfR.predict(data_test_selected)
prec_logR = average_precision_score(y_test, y_pred_logR)
acc_logR = accuracy_score(y_test, y_pred_logR)
classr_logR =classification_report(y_test, y_pred_logRq)

print(y_pred_logR)
print(prec_logR)
print(classr_logR)

OUTPUT
[1 0 0 ... 0 1 1]
0.09225134758511697
              precision    recall  f1-score   support

           0       0.91      0.58      0.71     18386
           1       0.09      0.43      0.15      1853

    accuracy                           0.56     20239
   macro avg       0.50      0.50      0.43     20239
weighted avg       0.83      0.56      0.66     20239



In [ ]:
#decion tree using SMote+RFECV data
DTR = DecisionTreeClassifier()
DTR.fit(x_train_smR, y_train_smR)

y_pred_DTR = DTR.predict(data_test_selected)
prec_DTR= average_precision_score(y_test, y_pred_DTR)
acc_DTR = accuracy_score(y_test, y_pred_DTR)
classr_DTR =classification_report(y_test, y_pred_DTR)

print(prec_DTR)
print(acc_DTR)
print(classr_DTR)

OUPUT
0.09228421888527652
0.681654231928455
              precision    recall  f1-score   support

           0       0.91      0.72      0.80     18386
           1       0.09      0.29      0.14      1853

    accuracy                           0.68     20239
   macro avg       0.50      0.50      0.47     20239
weighted avg       0.83      0.68      0.74     20239



In [ ]:
#Random forest using SMOTE +RFECV data

RFCR = RandomForestClassifier()
RFCR.fit(x_train_smR, y_train_smR)



y_pred_RFCR = RFCR.predict(data_test_selected)
prec_RFCR = average_precision_score(y_test, y_pred_RFCR)
acc_RFCR = accuracy_score(y_test, y_pred_RFCR)
classr_RFCR =classification_report(y_test, y_pred_RFCR)
print(prec_RFCR)
print(acc_RFCR)
print(classr_RFCR)

OUTPUT
0.0934171182554645
0.7998912989772222
              precision    recall  f1-score   support

           0       0.91      0.86      0.89     18386
           1       0.10      0.15      0.12      1853

    accuracy                           0.80     20239
   macro avg       0.51      0.51      0.51     20239
weighted avg       0.84      0.80      0.82     20239

